Add provider block reference rewrite (#557)
hashicorp/web-unified-docs: content/terraform/v1.12.x/docs/language/block/provider.mdx

---
page_title: Provider block reference for the Terraform configuration language
description: Use the `provider` block to declare and configure Terraform plugins, called providers. Providers let Terraform manage real-world infrastructure with provider-defined resources and data sources.
---

# `provider` block reference

Use the `provider` block to declare and configure Terraform plugins, called providers. Providers let Terraform manage real-world infrastructure with provider-defined resources and data sources.

> > **Hands-on:** Try the [Perform CRUD Operations with Providers](/terraform/tutorials/configuration-language/provider-use) tutorial.

Author: Good idea! I'm going to leave it here for now until we align because I've added tutorial call out to the other reference pages.

## Background

The `provider` block configures a named provider, which is a plugin that lets Terraform interact with cloud providers, SaaS providers, and other APIs.

HashiCorp's public [Terraform registry](https://registry.terraform.io/browse/providers) distributes providers separately from Terraform itself, and each provider has its own release cadence, documentation, and versions. If you are using HCP Terraform, you can use the [private registry](/terraform/cloud-docs/registry) to share providers within an organization.

Anyone can develop a Terraform provider and use it locally, or publish it to the Terraform public registry or HCP Terraform private registry. To learn more about authoring providers, refer to the [Plugin framework](/terraform/plugin/framework).

> Each Terraform module must declare which providers it requires, so that Terraform can install and use them. To declare a provider in a module, perform the following three steps:

Reviewer: ```suggestion
Each Terraform root module should declare which providers it requires, so that Terraform can install and use them. To declare a provider in a module, perform the following three steps:
```

So `root` module because per [our module docs](https://developer.hashicorp.com/terraform/language/modules/develop/providers):
```
Provider configurations, unlike most other concepts in Terraform, are global to an entire Terraform configuration and can be shared across module boundaries. Provider configurations can be defined only in a root Terraform module.
```

We mention this below (line 30), so maybe the last sentence can be left off here.

And `should` instead of `must` because Terraform will automagically install a matching provider based on resource name even if you don't add the `required_providers` or `provider` block...we should maybe mention this somewhere?

eg, this configuration by itself will work:

**main.tf**:
```
resource "random_pet" "foo" {
  length = 2
}

output "pet_name" {
  value = random_pet.foo.id
}
```

$ terraform init && terraform apply -auto-approve:
```
Initializing the backend...
Initializing provider plugins...
- Finding latest version of hashicorp/random...
- Installing hashicorp/random v3.7.2...
- Installed hashicorp/random v3.7.2 (signed by HashiCorp)
Terraform has created a lock file .terraform.lock.hcl to record the provider
## ...
random_pet.foo: Creation complete after 0s [id=capital-raptor]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

pet_name = "capital-raptor"
```

1. Define the provider's source, local name, and version in the [`required_providers` block](/terraform/language/block/terraform#required_providers) inside your top-level `terraform` block.
1. Add a top-level `provider` block to your configuration to configure a provider with authentication, a region, and other provider-specific arguments.
1. Install your provider.

>    - Running `terraform init` locally installs a provider and updates the [Dependency lock file](/terraform/language/files/dependency-lock) with the version you specified in your configuration.

Reviewer: ```suggestion
   - Running `terraform init` locally installs a provider and updates the [Dependency lock file](/terraform/language/files/dependency-lock) with the latest version matching the version string you configured in your `required_providers` block.
```

>    - HCP Terraform and Terraform Enterprise install providers as part of every run.

Reviewer: ```suggestion
   - HCP Terraform and Terraform Enterprise install providers as part of every run, using the versions provided in the dependency lock file, if it exists.
```

> If you are running Terraform locally, your plan and apply operations use the provider versions specified in the dependency lock file. You can upgrade a provider version in your lock file by changing the value in your configuration and re-running `terraform init`. To learn more about declaring providers, refer to [Provider requirements](/terraform/language/providers/requirements).

Reviewer: ```suggestion
If you are running Terraform locally, your plan and apply operations use the provider versions specified in the dependency lock file. You can upgrade a provider version in your lock file by changing the version in the `required_providers` block and re-running `terraform init`. To learn more about declaring providers, refer to [Provider requirements](/terraform/language/providers/requirements).
```

### `provider "<PROVIDER_NAME>"`

The name given in the block header specifies the [local name](/terraform/language/providers/requirements#local-names) of the provider to configure.

The following arguments are supported in a `provider` block:

| Argument | Description | Type | Required? |
| --- | --- | --- | --- |
| Provider-specific arguments | Configuration arguments that the provider developer defined. | Various | Varies by provider |
| `alias` | A unique identifier for a specific provider configuration, letting you use multiple configurations for the same provider. | String | Optional |
| `version` | A version constraint for the provider. This argument is deprecated. | String | Optional |

### Provider-specific arguments

The body of the `provider` block contains configuration arguments for the provider, which are defined by the provider itself.

<CodeBlockConfig hideClipboard>

```hcl
provider "<PROVIDER_NAME>" {
  <PROVIDER_ARGUMENTS>
}
```

</CodeBlockConfig>

For example, both the `project` and `region` arguments are defined by the `google` provider.

<CodeBlockConfig hideClipboard>

```hcl
provider "google" {
  project = "acme-app"
  region  = "us-central1"
}
```

> </CodeBlockConfig>

Author: I think having a real life example here is more helpful for understanding what is happening here, but I understand isn't doesn't follow the template. I'll remove!

A provider's documentation lists which configuration arguments it expects. For providers distributed on the [Terraform registry](https://registry.terraform.io), versioned documentation is available on each provider's page.

If you do not explicitly define a `provider` block, Terraform assumes and creates an empty default configuration for that provider. However, if a provider has required arguments, Terraform raises an error because it can't create that provider without the required values.

You can use [expressions](/terraform/language/expressions) to configure provider arguments, but you can only reference values that Terraform knows before it applies your configuration. Meaning, you can reference input variables and arguments that you specify directly in your configuration, but not any computed resource attributes, such as `google.web.public_ip`.

> Many providers support shell environment variables or other alternate sources for their configuration values, which helps keep credentials out of your version-controlled Terraform code. Refer to a provider's documentation to learn more about the assignment methods each argument supports.

Reviewer: ```suggestion
Many providers support shell environment variables or other alternate sources for their configuration values, which helps keep credentials out of your version-controlled Terraform configuration. Refer to a provider's documentation to learn more about the assignment methods each argument supports.
```

}
```

</CodeBlockConfig>

To refer to a provider alias, use `<PROVIDER_NAME>.<ALIAS>` in the `provider` argument of the following blocks:

- [`resource` blocks](/terraform/language/block/resource)
- [`data` blocks](/terraform/language/block/data)
- [`module` blocks](/terraform/language/modules/syntax)

If there are multiple aliases for a provider, the `provider` block without an `alias` argument is the default configuration for that provider. Resources, data sources, and modules that don't specify the [`provider` meta-argument](/terraform/language/block/resource#provider) use the default provider configuration that matches the resource type name.

#### Summary

- Data type: String
- Default: None
- Refer to the following examples to learn more about aliasing providers:
  - [Multiple provider configurations](#multiple-provider-configurations)
  - [Using an alternate provider configuration](#using-an-alternate-provider-configuration)
  - [Provider without a default configuration](#provider-without-a-default-configuration)
  - [Using an alternate provider configuration in a child module](#using-an-alternate-provider-configuration-in-a-child-module)

### `version`

> The `version` argument in provider configurations is deprecated, and Terraform will remove it in a future version. Instead, declare provider version constraints in [the `terraform` block's `required_providers` argument](/terraform/language/providers/requirements).

Reviewer: ```suggestion
The `version` argument in provider configurations is deprecated, and Terraform will remove it in a future version. Instead, declare provider version constraints in [the `terraform` block's `required_providers` block](/terraform/language/providers/requirements).
```

#### Summary

- Data type: String
- Default: None
- Deprecated

## Examples

The following examples demonstrate common use cases for `provider` blocks.

### Basic provider configuration

The following example configures the `google` provider with a specific project and region. In the `terraform` block's `required_providers` block you define the provider version you want to use, where to source the provider from, and the local name of the provider:

<CodeBlockConfig filename="terraform.tf">

```hcl
terraform {
  required_providers {
    google = {
      source  = "hashicorp/google"
      version = "~> 4.0"
    }
  }
}
```

</CodeBlockConfig>

> You can further configure a provider with the `provider` block in the root module of your configuration.

Reviewer: ```suggestion
You can configure the provider with the `provider` block in the root module of your configuration.
```

Maybe I'm splitting hairs, but my thinking is that the `required_providers` block configures *terraform* to use the specified provider, and the *provider's* configuration is in the provider block?

Reviewer: Also...should we mention somewhere that the `terraform { }` block being in `terraform.tf` and the `provider "google" { }` block in `main.tf` is our recommended convention?

Author: Oooh good point! I think the location of the blocks will be what we cover in the usage page though

<CodeBlockConfig filename="main.tf">

```hcl
provider "aws" {
  region = "us-east-1"
}

provider "aws" {
  alias  = "west"
  region = "us-west-2"
}
```

</CodeBlockConfig>

Resources that begin with `aws_` use the default `aws` provider configuration unless you supply the [`provider` argument](/terraform/language/block/resource#provider).

### Using an alternate provider configuration

You can make a resource, data source, or module use an alternate provider configuration by setting the `provider` meta-argument to a `<PROVIDER_NAME>.<ALIAS>` reference.

In the following example, the `aws_instance.foo` resource uses the `aws.west` provider configuration, and `aws_instance.bar` uses the default `aws` provider configuration:

<CodeBlockConfig highlight="6,11">

```hcl
provider "aws" {
  region = "us-east-1"
}

provider "aws" {
  alias  = "west"
  region = "us-west-2"
}

resource "aws_instance" "foo" {
  provider = aws.west
  # …
}

resource "aws_instance" "bar" {
   # …
}
```

</CodeBlockConfig>

### Provider without a default configuration

> If there are multiple aliases for a provider in your configuration, the `provider` block without an `alias` argument is the default configuration for that provider. If every `provider` block in your configuration uses an alias, Terraform creates an implied empty default configuration for that provider. Any resource that does not specify a `provider` meta-argument uses the empty default configuration.

Author: It's in there already! I just _also_ added it here to remind folk now that they have an example!

In the following example, both AWS provider configurations use aliases, but the `aws_s3_bucket` does not specify a `provider` argument:

```hcl
provider "aws" {
  alias  = "east"
  region = "us-east-1"
}

provider "aws" {
  alias  = "west"
  region = "us-west-2"
}

resource "aws_s3_bucket" "default_provider" {
  bucket = "uses-implied-empty-config"
}
```

Terraform uses the implied empty default configuration for the `aws_s3_bucket` resource because the resource doesn't specify a `provider` argument. If a provider requires specific arguments, Terraform returns an error when resources try to use the default configuration because the default is not properly configured.

### Pass provider configurations to a child module

> We recommend defining providers in root modules, and passing those provider configurations from parent modules to child modules to ensure your modules all use the same provider configurations. You can pass provider configurations from a root module to child modules using the [`providers` argument](/terraform/language/meta-arguments/module-providers).

Reviewer: This implies that we recommend explicitly passing providers to child modules rather than relying on implicit passing...I don't think we intend to do that, but I could be wrong...eg the main [use modules tutorial](https://developer.hashicorp.com/terraform/tutorials/modules/module-use#review-configuration) passes the provider implicitly.

Also, I'm not sure how much is recommendation vs. requirement, eg our [current docs say](https://developer.hashicorp.com/terraform/language/modules/develop/providers): `Provider configurations can be defined only in a root Terraform module.` - I don't think that's technically correct, but I think our intent is "Don't do this" rather than a soft recommendation.

Reviewer: Also also, assuming implicitly passing providers is ok, and the default behavior is fine, then we should introduce an example of that first, before explicit example.

    aws = {
      source  = "hashicorp/aws"
      version = "~> 5.0"
      configuration_aliases = [aws.west]
    }
  }
}

data "aws_ami" "amazon_linux" {
  provider = aws.west

  #...
}
```

</CodeBlockConfig>

Modules have special requirements for providers, refer to [Providers within modules](/language/modules/develop/providers) to learn more.

### Provider configuration with expressions

In the following example, the provider configuration uses input variables and local values to set its arguments:

<CodeBlockConfig highlight="15,18">

```hcl
variable "aws_region" {
  description = "The AWS region to deploy resources in"
  type        = string
  default     = "us-west-2"
}

locals {
  common_tags = {
    Environment = "production"
    Project     = "web-app"
  }
}

provider "aws" {
  region = var.aws_region

  default_tags {
    tags = local.common_tags
  }
}
```

</CodeBlockConfig>

### Provider configuration without arguments

In the following example, the provider block defines an empty configuration for the `random` provider. Terraform also assumes an empty default configuration for any provider that you do not explicitly configure with a `provider` block:

> ```hcl
> provider "random" {
> 
> }
> ```

Reviewer: ```suggestion
```hcl
provider "random" { }
```
```

I see this style in examples more often than the line-break style for "empty" provider configuration.
